[OMID-72] Add fences to support secondary index in Phoenix

ohadshacham opened the pull request:
This commit adds support for fences over a table. These fences guarantee that every transaction started before the fence and writes to the table the fence was created for will be aborted.
These fences are needed for creating a secondary index in Apache Phoenix. The scenario was discussed in TEPHRA-157 and PHOENIX-2478. Augmenting Omid with this kind of transactions was also discussed in OMID-56.

francisco-perez-sorrosal committed:
[OMID-64] Fix location of protobuf tgz in travis config

Location has changed to
https://github.com/google/protobuf/releases/download/v2.5.0/protobuf-2.5.0.tar.gz

Change-Id: Icd7740ff23bf17142fb47bdc530178b0292b3a05

Commit:
This commit is related to jira [OMID-72].

This commit adds support for fences over a table. These fences guarantee that every transaction started before the fence and writes to the table the fence was created for will be aborted.
These fences are needed for creating a secondary index in Apache Phoenix. The scenario was discussed in TEPHRA-157 and PHOENIX-2478. Augmenting Omid with this kind of transactions was also discussed in OMID-56.

ebortnik commented on transaction-client/src/main/java/org/apache/omid/transaction/AbstractTransactionManager.java:
```diff
@@ -159,6 +163,30 @@ public final Transaction begin() throws TransactionException {
+    @Override
+    public final Transaction fence(byte[] tableName) throws TransactionException {
+        long fenceTimestamp;
+        long tableID = Hashing.murmur3_128().newHasher().putBytes(tableName).hash().asLong();
```
Copy&paste from HBaseCellId.getTableId(). Please reuse code.

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/PersistenceProcessorImpl.java:
```diff
@@ -146,20 +146,30 @@ public void addCommitRetryToBatch(long startTimestamp, Channel c, MonitoringCont
 
-        currentBatch.addTimestamp(startTimestamp, c, context);
+        currentBatch.addTimestamp(startTimestamp, c, monCtx);
+        if (currentBatch.isFull()) {
```
Where did this code reside before? Who was responsible for triggering the flush?

ohadshacham replied:
Every function that adds an item to a batch should check whether the batch is full and trigger a flush in case it is. You can take a look at add{Timestamp, Abort, Commit, CommitRetry}ToBatch functions. We can wrap this check in some epilog.

ebortnik replied:
Yes, it would be good to have all addXXX methods go thru the same funnel and flush at the end.

ebortnik commented on transaction-client/src/main/java/org/apache/omid/tso/client/MockTSOClient.java:
```diff
@@ -46,6 +51,23 @@
     }
 
     @Override
+    public TSOFuture<Long> getFence(long tableId) {
```
Please select the right location for a comment about how you you model a fence (as a txn with begin ts = commit ts).

ohadshacham replied:
The returned transaction contains the fence timestamp as the transaction ID, documented at TransactionManager interface.
The TSO persists the fence by writing to the commit table the fence timestamp as both the start and commit timestamp, documented at PersistenceProcessorHandler.

ebortnik commented on transaction-client/src/main/java/org/apache/omid/tso/client/MockTSOClient.java:
```diff
@@ -46,6 +51,23 @@
+            f.set(fenceTimestamp);
+            fenceMap.put(tableId, fenceTimestamp);
+            try {
+                commitTable.addCommittedTransaction(fenceTimestamp, fenceTimestamp);
```
Same comment, here

ebortnik commented on transaction-client/src/main/java/org/apache/omid/tso/client/MockTSOClient.java:
```diff
@@ -54,13 +76,31 @@
+                tableIDs.add(c.getTableId());
+            }
+
+            boolean canCommit = true;
```
Please add minimal comments to this code. Also, the method is long. Suggest to split it into helper functions, i.e., two boolean methods hasConflictsWithFences(), hasConflictsWithCommittedTransactions()

ebortnik commented on transaction-client/src/main/java/org/apache/omid/tso/client/TSOClient.java:
```diff
@@ -92,6 +95,8 @@
     private InetSocketAddress tsoAddr;
     private String zkCurrentTsoPath;
 
+    private final Set<Long> tableIDs;
```
Please add comment why you need this.

ebortnik commented on transaction-client/src/main/java/org/apache/omid/tso/client/TSOClient.java:
```diff
@@ -626,6 +664,7 @@ public void run(Timeout timeout) {
+            fenceRequests = new HashMap<>();
         }
 
         private Timeout newTimeout(final StateMachine.Event timeoutEvent) {
```
Huge method, consider splitting ..

ebortnik commented on transaction-client/src/main/java/org/apache/omid/tso/client/TSOClient.java:
```diff
@@ -762,6 +829,15 @@ private void handleError(StateMachine.Fsm fsm) {
                 queueRetryOrError(fsm, r.getRequest());
                 iter.remove();
             }
+            iter = fenceRequests.entrySet().iterator();
```
Can you do a more modern-style iteration?

ohadshacham replied:
I kept the same syntax used at the lines above. I prefer to either keep the same syntax or refactor the function. We can do another refactoring commit.

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/PersistenceProcessor.java:
```diff
@@ -33,6 +33,8 @@ void addCommitToBatch(long startTimestamp, long commitTimestamp, Channel c, Moni
 
     void addTimestampToBatch(long startTimestamp, Channel c, MonitoringContext monCtx) throws Exception;
 
+    void addFenceToBatch(long tableID, long fenceTimestamp, Channel c, MonitoringContext monCtx) throws Exception;
```
Here too, please add a comment about how a fence is modeled.

ohadshacham replied:
Added at the PersistenceProcessorHandler.

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/RequestProcessor.java:
```diff
@@ -27,6 +27,7 @@
     void timestampRequest(Channel c, MonitoringContext monCtx);
 
-    void commitRequest(long startTimestamp, Collection<Long> writeSet, boolean isRetry, Channel c, MonitoringContext monCtx);
+    void commitRequest(long startTimestamp, Collection<Long> writeSet, Collection<Long> TableIdSet, boolean isRetry, Channel c, MonitoringContext monCtx);
```
TableIdSet --> tableIdSet

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/RequestProcessorImpl.java:
```diff
@@ -147,13 +156,24 @@ public void timestampRequest(Channel c, MonitoringContext monCtx) {
 
     @Override
-    public void commitRequest(long startTimestamp, Collection<Long> writeSet, boolean isRetry, Channel c,
+    public void commitRequest(long startTimestamp, Collection<Long> writeSet, Collection<Long> TableIdSet, boolean isRetry, Channel c,
```
TableIdSet --> tableIdSet

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/RequestProcessorImpl.java:
```diff
@@ -180,20 +201,35 @@ private void handleCommit(RequestEvent event) throws Exception {
-                if (value != 0 && value >= startTimestamp) {
-                    txCanCommit = false;
-                    break;
+            // 1. check table fences
```
Please split the method. Each of (1), (2) and (3) could be a function.

Commit:
Integrating ebortnik comments to pull request #125

Commit:
Merge branch 'master' into DDLandDMLfences

Commit:
[OMID-71] Omid's transaction manager detects conflicts at the
 cell level. Meaning that two concurrent transactions write to different
 columns at the same row do not conflict. This semantics is not suitable for
 Apache Phoenix which requires conflict detection at the row level. This
 commit augments Omid with row level conflict analysis.

Commit:
[OMID-70]  Adding world base timestamp generator. The
 generator maintains monotonicity and persists the maximal timestamp in each
 interval. The motivation for this commit is the integration with Apache
 Phoenix. Phoenix's customers requires mapping from timestamp to world time in
 order to debug their logs. This commit also contains augmentation of the
 proto in order to partially serialize transaction. This is also for Phoenix
 usage.

Commit:
Merge branch 'master' into DDLandDMLfences
